### PR TITLE
OCS S3 Snapstore support access information via mounted secret

### PR DESCRIPTION
**Pull request by Wieneo**
**What this PR does / why we need it**
OCS S3 snapstore now supports the new way of supplying access information via a secret mounted to a specific directory.
This allows for copying ETCd backups between OCS buckets with etcd-druid.

**Release note**:

```improvement operator
OCS S3 Snapstore now supports supplying access information via a mounted secret.
```

### CODE REVIEW

**Comment by gardener-robot-ci-1** [comment]
Thank you @Wieneo for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

File: pkg/snapstore/ocs_s3_snapstore.go [hunk 1]
```diff
@@ -28,32 +41,166 @@ const (
...
+	files, err := ioutil.ReadDir(dirname)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		switch file.Name() {
+		case "endpoint":
+			{
+				data, err := ioutil.ReadFile(dirname + "/endpoint")
```
Comment (ishan16696) [inline pkg/snapstore/ocs_s3_snapstore.go:108]:
```suggestion
				data, err := os.ReadFile(dirname + "/endpoint")
```
 see this https://go.dev/doc/go1.16#ioutil

File: pkg/snapstore/ocs_s3_snapstore.go [hunk 2]
```diff
@@ -28,32 +41,166 @@ const (
...
+				if err != nil {
+					return nil, err
+				}
+				ao.Endpoint = string(data)
+			}
+		case "accessKeyID":
+			{
+				data, err := ioutil.ReadFile(dirname + "/accessKeyID")
```
Comment (ishan16696) [inline pkg/snapstore/ocs_s3_snapstore.go:116]:
```suggestion
				data, err := os.ReadFile(dirname + "/accessKeyID")
```
same as above and for all places

**Review by ishan16696 — CHANGES_REQUESTED** [review]
Thanks @Wieneo for your contribution.
Please also update the [document](https://github.com/gardener/etcd-backup-restore/blob/master/doc/usage/getting_started.md#various-ways-to-pass-credentials) and add secret example https://github.com/gardener/etcd-backup-restore/tree/master/example/storage-provider-secrets

**Comment by Wieneo** [comment]
@ishan16696 I've added the requested docs / examples :)

**Comment by Wieneo** [comment]
@ishan16696 Is there still something missing? :)

**Comment by ishan16696** [comment]
>Is there still something missing? :)

Can you please also address the review comments. Thanks

**Comment by Wieneo** [comment]
Ah, okay :) I didn't realize I had to accept your proposed commits.
Should be fine now @ishan16696

**Comment by ishan16696** [comment]
>I didn't realize I had to accept your proposed commits.

no, you don’t have to. You can have a counter argument(if any). We will go with that change which is best for PR and code.

File: pkg/snapstore/ocs_s3_snapstore.go [hunk 3]
```diff
@@ -28,32 +41,166 @@ const (
...
+					return nil, err
+				}
+				ao.DisableSSL, err = strconv.ParseBool(string(data))
+				if err != nil {
+					return nil, err
+				}
+			}
+		case "insecureSkipVerify":
+			{
+				data, err := ioutil.ReadFile(dirname + "/insecureSkipVerify")
```
Comment (ishan16696) [inline pkg/snapstore/ocs_s3_snapstore.go:151]: same as above

**Review by ishan16696 — CHANGES_REQUESTED** [review]
Please address these changes.

**Comment by Wieneo** [comment]
I didn't know about https://go.dev/doc/go1.16#ioutil, thank you! :)
I know I could have a counter argument, I just don't have very much experience with GitHub PR's and didn't realize I had to do something. 

I've replaced the ioutil with the os package.

**Comment by ishan16696** [comment]
thanks @Wieneo for addressing my review comments.
One more change required:
Basically if you check [here](https://github.com/gardener/etcd-backup-restore/blob/64eb001c2021b3a0840a0e326c2e9ddcca86785b/pkg/snapshot/snapshotter/snapshotter.go#L420-L431) , here we are checking whether a secrets updated or not by calculating the hash of secrets.
So, can you also add a hash function [here](https://github.com/gardener/etcd-backup-restore/blob/64eb001c2021b3a0840a0e326c2e9ddcca86785b/pkg/snapstore/utils.go#L175-L192) for `OCS`.

**Comment by Wieneo** [comment]
Hi @ishan16696!
I've added the requested hash function to the OCS snapstore provider.

File: pkg/snapstore/ocs_s3_snapstore.go [hunk 4]
```diff
@@ -64,14 +210,46 @@ func ocsAuthOptionsFromEnv() (s3AuthOptions, error) {
...
+	ao, err := getOCSAuthOptions("")
+	if err != nil {
+		return "", err
+	}
+
+	return getOCSHash(ao), nil
+}
+
+func getOCSHash(config *ocsAuthOptions) string {
+	data := fmt.Sprintf("%s%s%s%s%v%v", config.AccessKeyID, config.Endpoint, config.Region, config.SecretAccessKey, config.DisableSSL, config.InsecureSkipVerify)
```
Comment (ishan16696) [inline pkg/snapstore/ocs_s3_snapstore.go:253]: `config.DisableSSL`, `config.InsecureSkipVerify` can we remove these 2 as they are optional ?

**Review by ishan16696 — COMMENTED** [review]
Overall looks good, Only one doubt:

File: pkg/snapstore/ocs_s3_snapstore.go (hunk 4 above)
Comment (Wieneo) [reply pkg/snapstore/ocs_s3_snapstore.go]: You are right :)
I've removed them from the hashing function.

**Comment by ishan16696** [comment]
One check in pipeline got failed with this error:
`exported function OCSSnapStoreHash should have comment`. Please add a comment. You can also check this locally by running `make check`

**Comment by Wieneo** [comment]
Added a comment to the mentioned function.
`make check` completed without a hitch.

**Comment by ishan16696** [comment]
Please sign the CLA.

**Comment by Wieneo** [comment]
I signed the CLA using the link https://cla-assistant.io/gardener/etcd-backup-restore?pullRequest=465

**Comment by ishan16696** [comment]
CLA status in pipeline is not getting updated even after I have re-run the tests. 
Kindly asking you to do a force-push @Wieneo .

**Comment by Wieneo** [comment]
I've pushed a blank commit.
The CLA check now seems happy :)

**Review by ishan16696 — APPROVED** [review]
LGTM!!

**Pull request merged**